From 9235ffd2f0d5cf80b73b9a30226d034c37e7239c Mon Sep 17 00:00:00 2001
From: Huang Chaofan <chaofan.huang@nxp.com>
Date: Mon, 2 Apr 2018 18:42:37 +0800
Subject: [PATCH 3557/5242] MLK-17967[IMX8QXP B0] Judge the buffer status when
 queue buffer

commit  7fcbc5701577709c666ee73988fe51d6b1b89ae4 from
https://source.codeaurora.org/external/imx/linux-imx.git

Judge the buffer status when it is free or decoded, that means it is
used by the firmware, don't put it on the list

Signed-off-by: Huang Chaofan <chaofan.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-decoder-b0/vpu_b0.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index f836063..99784db 100755
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -806,6 +806,7 @@ static int v4l2_ioctl_streamoff(struct file *file,
 		ctx->wait_rst_done = true;
 		v4l2_vpu_send_cmd(ctx, ctx->str_index, VID_API_CMD_ABORT, 0, NULL);
 		add_eos(ctx, 0);
+		wake_up_interruptible(&ctx->buffer_wq);
 		wait_for_completion(&ctx->completion);
 	}
 
@@ -1855,7 +1856,8 @@ static void vpu_buf_queue(struct vb2_buffer *vb)
 	data_req->vb2_buf = vb;
 	data_req->id = vb->index;
 
-	list_add_tail(&data_req->list, &This->drv_q);
+	if (data_req->status != FRAME_FREE && data_req->status != FRAME_DECODED)
+		list_add_tail(&data_req->list, &This->drv_q);
 
 	vpu_dbg(LVL_INFO, "before c_port_buf_queue up, vq->type=%d, vb->index=%d\n", vq->type, vb->index);
 	up(&This->drv_q_lock);
-- 
1.7.9.5

